
<u>Recall</u>:
```c++
ostream &operator<<(ostream &out, const list &l) {
	for (...) {
		out << *it << ' ';
		++*it; // increment the items as you print
	}
	return out
}
```
<u>will</u> compile. Shouldn't - the list is supposed to be const.
compiles because operator * produces a non-const ref.
<u>Should</u> compile if l is non-const.


<u>Conclusion</u> - iteration over const is different from iteration over non-const
-Make a second iterator class.

```c++
class list {
	...
	public:
		class iterator {
			...
			public:
				int &operator*();
				...
		};

		class const_iterator {
			Node *p;
			public:
				const_iterator(Node *p): p{p};
				bool operator!=(const const_iterator other) const {
					return p != other.p;
				}
				
				const_iterator& operator++() {
					p = p->next;
					return *this;
				}
				
				const int& operator*() const {
					return p->data;
				}
		};
		
		iterator begin() {return iterator{theList};}
		iterator end() {return iterator{nullptr};}
		
		const_iterator begin() const {
			return const_iterator{theList};
		}
		
		const_iterator end() const {
			return const_iterator{nullptr};
		}
};
```


```cpp
ostream &operator<<(ostream &out, const list &l) {
	for (list::const_iterator it = l.begin(); it != l.end(); ++it) {
		out << *it << ' ';
	}
	return out;
}
```
- Works now;
- `list::const_iterator it` - mouthful
- Shorter: `auto it`

`auto x = expr` - saves writing down x's type
-x will be given the same type as expr's value.


Shorter cut:
```c++
ostream &operator<<(ostream *out, const list &l) {
	// range-based for loop
	for (auto n : l) {
		out << n << ' ';
	}
	return out;
}
```
Ranged-based for loops - available for any class with:
- methods (or functions) begin+end that return an iterator obj.
- the iterator class must support unary \*, prefix++, operator !=


<u>Note:</u>
```c++
// n declared by value: ++n increments n, not the list items;
for (auto n : l) ++n; 

// n is a ref - will update list elements
for (auto &n : l) ++n;

// n by ref (no copying) - cannot be mutated
for (const auto &n : l)
```



One small encapsulation problem ...
<u>client</u>: `list::iterator it{nullptr};` 
- forgery - can create an end iterator without calling l.end().

<u>To fix:</u> 
make iterator constructors <u>private</u>. Then the client can't call them.

<u>But then:</u>
`list` can't create them either.

<u>Solution:</u> *Friendship*
```c++
class list {
	...
	public: 
		class iterator {
			...
			iterator(Node *p);
			public:
				...
				friend class list;
		};
		
		class const_iterator {
			...
			const_iterator(Node *p);
			public:
				...
				friend class list;
		};
};
```
Meaning now list has access to <u>all</u> of iterator's/const_iterator's implementation.


<u>Limit friendships.</u> - they weaken encapsulation.


Can do the same for vectors.
```c++
class vector {
	int *theVector;
	size_t n, cap;
	
	public:
		class iterator {
			int *p;
		};
		
		class const_iterator {
			const int *p;
		};

		// OR
		typedef int* iterator;
		typedef const int* const_iterator;

		// OR
		using iterator = int*;
		using const_iterator = const int*;


		iterator begin() {return theVector;}
		const_iterator begin() const {return theVector;}

		iterator end() { return theVector + n;}
		const_iterator end() const {return theVector + n;}
};
```




# Problem 11: Now you've gone too far!

```c++
vector v;
v.push_back(4);
v.push_back(5);
v[2]; // out of bounds (UB - may or may not crash)
```

Can we make this safer?

Adding bounds check to operator\[] - may be needlessly expensive
Could have a second, safer, fetch operator.
```c++
class vector {
	...
	public:
		int &at(size_t i) {
			if (i < n) {
				return theVector[i]
			}
			else //what? ...
			// v.at(2) still wrong - what should happen?
		}
}
```
returning any int - looks like a non-error
returning a non-int - not type-correct
crash the program? - Can we do better?
-> Don't return, don't crash.

<u>Solution:</u> Raise an *exception*
```c++
class range_error {};
class vector {
	int &at (size_t i) {
		if (i < n) {...}
		else throw range_error{};
		// construct an object of type range_error and "throw" it.
	}
}
```

<u>Client's options:</u>
1. Do nothing - 
	```c++
	vector v;
	v.push_back(1);
	v.at(1); // the exception will crash the program.
	```
2.  catch it
```c++
	try {
		vector v;
		v.at(0);
	}
	// r is the thrown object
	catch (range_error &r) {
		// do something
	}
```
3. let your caller catch it.
